editors/code/src/commands/inlay_hints.ts

```diff
@@ -21,32 +21,44 @@ const typeHintDecorationType = vscode.window.createTextEditorDecorationType({
 
 export class HintsUpdater {
     private displayHints = true;
+    private drawnDecorations = new Map<string, vscode.DecorationOptions[]>();
 
-    public async loadHints(
-        editor: vscode.TextEditor | undefined
-    ): Promise<void> {
-        if (
-            this.displayHints &&
-            editor !== undefined &&
-            this.isRustDocument(editor.document)
-        ) {
-            await this.updateDecorationsFromServer(
-                editor.document.uri.toString(),
-                editor
-            );
+    public async loadHints(editor?: vscode.TextEditor): Promise<void> {
+        if (this.displayHints) {
+            const documentUri = this.getEditorDocumentUri(editor);
+            if (documentUri !== null) {
+                const latestDecorations = this.drawnDecorations.get(
+                    documentUri
+                );
+                if (latestDecorations === undefined) {
+                    await this.updateDecorationsFromServer(
+                        documentUri,
+                        editor!
+                    );
+                } else {
+                    await editor!.setDecorations(
```

Comment:
What if the file changed outside of the control of vscode?

Reply:
On my machine (macOS) vscode is fine with tracking external file changes in `rust-analyzer` project: it emits `vscode.TextDocumentChangeEvent` which we handle via the `updateHints` method.

I've also found some old issue that hints that vscode devs try to support the file watch as better as possible: https://github.com/microsoft/vscode/issues/31600

If vscode fails to track the external file change at some point, we mess up code coloring and current inlay hints from `master` branch anyway, so my PR does not make it worse :)

Reply:
Yeah, I think VS Code should work strictly in terms of in-memory representation of the file. If in memory and on-disk contents are different and an extension observes it, than that's the bug in VS Code.

Reply:
> so my PR does not make it worse

That's actually incorrect: `master` branch would have made a server request and updated the hints, while this version displays incorrect old hints from `drawnDecorations`.

Luckily, the new hints will be queried after the document is changed in the vscode, so it is easy to override.

I think this kind of situations should not be frequent in real life and a fair price for not querying the server for every `.rs` file opened in the editor (as we do it now)

Reply:
>I think this kind of situations should not be frequent in real life and a fair price for not querying the server for every .rs file opened in the editor (as we do it now)

We actually *should* query the server for each of .rs files: change in one file can affect type-hints in another file. If we want to save computation here, we should add an ability to filter type-hints by the viewport (so that we query only visible things).

Reply:
Indeed, how could I miss it?

What if we still keep the cache and clean it when the next edit is made?
Decorations disappear every time the editor is closed, but we might reuse the ones already queried if no changes are made?